[torch] Add backward support for segment reduce (CPU only)

serhaty opened the pull request:
Summary:
This is to setup boiler plate code for backward and CPU implementation.

Next Steps in order:
- Add backward support for CUDA
- Add support for more aggregation types
- Benchmarking (for cuda mainly)/more testing/documentation
- Support for multi dimension

Test Plan:
Updated unit test to also check correctness of backward.

Wait for CI signal

Differential Revision: D27970340

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8945,4 +8945,9 @@
   variants: function
   dispatch:
-    CPU: _segment_reduce_cpu
+    CPU, CUDA: segment_reduce_kernel
```
This function looks pretty fresh. Consider adding an out= variant and making it structured, see https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -60,6 +61,54 @@ Tensor _segment_reduce_cpu_kernel(
+    const Tensor& output,
+    const Tensor& data,
+    const Tensor& lengths) {
+  auto grad_output = at::zeros({data.sizes()}, grad.options());
```
This variable should be called `grad_input` (it is gradient with respect to operation input), `grad_output` or `grad` is the gradient you receive

serhaty replied:
Sounds good, updated.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -60,6 +61,54 @@ Tensor _segment_reduce_cpu_kernel(
+          for (int64_t j = 0; j < lengths_data[i]; ++j) {
+            if (at::_isnan(values_data[k]) ||
+                values_data[k] == output_data[i]) {
+              grad_output_data[k] = 1;
```
`grad_output_data[k] = grad[i]`, and compare `grad_output_data[index] !=0` in the next loop. Also, as pointed above, the name should be `grad_input`.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -60,6 +61,54 @@ Tensor _segment_reduce_cpu_kernel(
+            k++;
+          }
+          // Average gradient output based on number of maximum in the segment
+          TORCH_CHECK(counter > 0);
```
This should be TORCH_INTERNAL_ASSERT

ngimel commented on test/test_segment_reductions.py:
```diff
@@ -11,28 +11,40 @@
+        actual_result.sum().backward()
         self.assertEqual(
-            expected_result, actual_result, rtol=1e-03, atol=1e-05, equal_nan=True
+            expected_grad, data.grad, rtol=1e-03, atol=1e-05, equal_nan=True
```
direct checking of gradients is good, but you should also test with `gradcheck`

serhaty replied:
is there an example I can check for this that works with ops that have key value arguments?

Commit:
[torch] Add backward support for segment reduce (CPU only)

Summary:
This is to setup boiler plate code for backward and CPU implementation.

Next Steps in order:
- Add backward support for CUDA
- Add support for more aggregation types
- Benchmarking (for cuda mainly)/more testing/documentation
- Support for multi dimension

Test Plan:
Updated unit test to also check correctness of backward.

Wait for CI signal

Differential Revision: D27970340

fbshipit-source-id: c961dbef03678fa5506ae30b3ffe15334b7f3a4e

ngimel commented on test/test_segment_reductions.py:
```diff
@@ -7,32 +7,64 @@
 
+        # gradcheck does not work well with bfloat16 or fp16 cpu types
+        # also there is small numerical difference with fp32
+        if dtype not in [torch.half, torch.bfloat16, torch.float]:
```
check grad only for double (`if dtype == torch.double`)

ngimel reviewed (approved):
Looks good, please consider making it structured in the followup, as @ezyang suggests

serhaty commented:
This PS has been checked-in. Closing it manually.
